internal/provider/datasources/deployment.go

```diff
@@ -0,0 +1,405 @@
+package datasources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-framework-jsontypes/jsontypes"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/prefecthq/terraform-provider-prefect/internal/api"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/customtypes"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/helpers"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/resources"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &deploymentDataSource{}
+	_ datasource.DataSourceWithConfigure = &deploymentDataSource{}
+)
+
+// deploymentDataSource is the data source implementation.
+type deploymentDataSource struct {
+	client api.PrefectClient
+}
+
+// DeploymentDataSourceModel defines the Terraform data source model.
+type DeploymentDataSourceModel struct {
+	// The model requires the same fields, so reuse the fields defined in the resource model.
+	resources.DeploymentResourceModel
+}
```

Comment:
afaik, this is the first time trying something like this - let me know if there are any drawbacks I may have missed

Reply:
honestly, im not sure there is a drawback yet. they use the same `types` package

Reply:
was thinking we could make a new `common` package next to `resources` and `datasources` and put common elements there 🤔